bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -137,18 +136,18 @@ template <>
 void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequestMsg *msg) {
   if (preProcessorMetrics_.requestReceived.Get().Get() % 10 == 0) metricsComponent_.UpdateAggregator();
   preProcessorMetrics_.requestReceived.Get().Inc();
-  ClientPreProcessReqMsgUniquePtr clientPreProcessReqMsg(msg);
-  if (!clientPreProcessReqMsg) return;
 
-  const NodeIdType &senderId = clientPreProcessReqMsg->senderId();
-  const NodeIdType &clientId = clientPreProcessReqMsg->clientProxyId();
-  const ReqId &reqSeqNum = clientPreProcessReqMsg->requestSeqNum();
+  // Cannot wrap ClientPreProcessRequestMsg into a smart pointer, like other messages as its life cycle depends
+  // on pre-processing success.
+  const NodeIdType &senderId = msg->senderId();
+  const NodeIdType &clientId = msg->clientProxyId();
+  const ReqId &reqSeqNum = msg->requestSeqNum();
   LOG_INFO(GL,
            "Received ClientPreProcessRequestMsg reqSeqNum=" << reqSeqNum << " from clientId=" << clientId
                                                             << ", senderId=" << senderId);
-  if (!checkClientMsgCorrectness(clientPreProcessReqMsg, reqSeqNum)) {
+  if (!checkClientMsgCorrectness(msg, reqSeqNum)) {
     preProcessorMetrics_.requestIgnored.Get().Inc();
-    return;
+    return delete msg;
```

Comment:
Instead of using `delete`, can we make `msg` a smart pointer? This way we would be more resilient to future changes.

Reply:
No. It was a smart pointer before but in order to be able to control this pointer time life cycle in a smart way without unneeded data copies, I need to have it raw. I've added an explaining comment in the code.